Shell script to simplify upload, ingestion, etc.

andkaha opened the pull request:
Closes #2 

Note that the script's name, `sda`, is a "working title" and may change. EDIT: The script is now called `sda-admin`.

Slightly old and outdated help text:

```
General synopsis:
    ./sda [GLOBAL OPTIONS] [help] {upload|ingest|accession|dataset} [ARGUMENTS]

Global options:
    --sda-config pathname       SDA S3 configuration file       Default: s3cmd.conf
    --sda-key pathname          SDA CRYPT4GH public key file    Default: crypt4gh_key.pub

Specific synopsis:
    ./sda help

    ./sda [...] upload pathname [pathname...]
    ./sda help upload

    ./sda [...] ingest filename [filename...]
    ./sda [...] ingest
    ./sda help ingest

    ./sda [...] accession accessionID filename
    ./sda [...] accession
    ./sda help accession

    ./sda [...] dataset datasetID filename [filename...]
    ./sda [...] dataset
    ./sda help dataset
```

```
The "upload" sub-command is used for encrypting and uploading
one or several files to the configured S3 storage's inbox.
Both encryption and uploading is carried out using "sda-cli".

If a given file is already encrypted (it has a ".c4gh" filename
suffix), it will not be re-encrypted.  If a given file exists
alongside a corresponding encrypted file, the encrypted file
will be used in place of the given file.  To ensure that each
file is freshly encrypted, remove all encrypted versions of the
files.

Example usage:

    Files may be uploaded one by one or several at once.
    Encrypting and uploading three files (this creates
    "file1.c4gh", "file2.c4gh", and "file3.c4gh", if these files
    did not already exist):

    ./sda upload file1 file2
    ./sda upload file3
```

```
The "ingest" sub-command is used for ingesting one or several
uploaded files.  The filenames to be ingested may be given with
or without the ".c4gh" filename suffix and with or without a
full directory path (the suffix will be added to names that lack
it, and any directory path will be ignored).

Example usage:

    Listing the filenames currently in the "inbox" queue waiting
    to be ingested:

    ./sda ingest

    Files may be ingested one by one or several at once.
    Ingesting three files:

    ./sda ingest file1 file2
    ./sda ingest file3
```

```
The "accession" sub-command is used for assignig an accession ID
to a single file that has previously been ingested.  As with the
"ingest" sub-command, the filenames may be given with or without
the ".c4gh" filename suffix and with or without a full directory
path.

Example usage:

    Listing the filenames currently in the "verified" queue
    waiting to have accession IDs assigned to them:

    ./sda accession

    Accessions are only ever assiged to one file at a time.
    Assigning accessions to three files:

    ./sda accession MYID001 file1
    ./sda accession MYID002 file2
    ./sda accession MYID003 file3
```

```
The "dataset" sub-command is used for associating one or several
files to a single dataset ID.  As with the "ingest" sub-command,
the filenames may be given with or without the ".c4gh" filename
suffix and with or without a full directory path.

Example usage:

    Listing the filenames currently in the "completed" queue
    waiting to be associated with a dataset ID:

    ./sda dataset

    Files are associated to dataset IDs one at a time or several
    at once.  Associating three files with a dataset ID:

    ./sda dataset MYSET001 file1 file2
    ./sda dataset MYSET001 file3
```

andkaha committed:
WIP: Script to make upload, ingestion etc. from the command line simple

Note that the script's name, `sda`, is a "working title" and may change.
Run the script with no option to get help text.

Currently implements `sda upload` partially (will fail due to missing
certificates).

andkaha committed:
Merge branch 'main' into feature/helper-script

andkaha committed:
Streamline cert4gh installation and certificate generation

andkaha committed:
Help text update and cleanup

andkaha committed:
Make ShellCheck happy

Setting a variable to the value `test` (the name of a known standard
utility) makes ShellCheck think that you meant to use `$(test)`.  We
will use `testuser` instead.

andkaha committed:
Return early

andkaha committed:
Use `sda-cli`, so delete most of what I wrote yesterday

andkaha committed:
Encrypt all files in one go, then upload them all

... and use encypted version of files, if found.

andkaha committed:
Less handholding, upload done, work on ingestion

Ingestion now finds relevant correlation IDs:

```shell
% ./sda ingest file1 file hhh file
955d3253-cd8e-4cdc-9111-ee60e5a454cf
4d7f1860-850e-4bba-ac1a-6ef3360e7f7d
```

Above, the two files `file` and `file1` were previously uploaded
(actually, uploaded *multiple* times).

andkaha committed:
ACK all inbox messages, and later requeue the ones we're not interested in

andkaha committed:
Publish better

andkaha committed:
Ingestion probably works now

andkaha committed:
Don't remove things we don't need to remove

andkaha committed:
Clean up code, use base64 for payload

andkaha committed:
Neater URLs, and allow for listing ingestable files

andkaha committed:
Various

* Neater code.
* Use the file path from the message payload for deduplication.
* Allow the for showing what files are ready for getting accession IDs.

andkaha committed:
refactor

andkaha committed:
Accession code probably done, started on lat bit with datasets

andkaha committed:
Read the access key only when needed

andkaha committed:
Make the `publish` helper function quiet

andkaha committed:
Sort out some semantics and finish dataset implementation

What's left now:

* The help messages.
* Error handling?

andkaha committed:
Initial `upload` help text

andkaha committed:
Rearrange

andkaha committed:
Initial `ingest` usage text

andkaha committed:
Add initial `accession` usage text

andkaha committed:
Initial `dataset` usage text, and some smaller adjustments

andkaha committed:
Merge branch 'main' into feature/helper-script

andkaha committed:
Remove `set -x`

andkaha committed:
`tojson` not needed

andkaha committed:
Add missing comments, clarify old comments

andkaha committed:
Rename script from `sda` to `sda-admin`

andkaha committed:
Consistent use of short-circuit test

andkaha committed:
Make MQ host/credentials settable via command line options

jbygdell commented on scripts/sda-admin:
```diff
@@ -0,0 +1,448 @@
+
+# Default values for global options.
+mq_credentials=test:test	# --mq-credentials user:pass
+mq_host=localhost		# --mq-host hostname
```
If `mq_host` is the entire URL i.e. `http(s)://hostname:port` then you don't have to modify the script for a semi-production environment where TLS is enabled.

andkaha replied:
> If `mq_host` is the entire URL i.e. `http(s)://hostname:port` then you don't have to modify the script for a semi-production environment where TLS is enabled.

Should I also rename the variable `mq_url` or similar?

jbygdell replied:
Yes, since this script will be used by 3:rd party people we need to make it as clear as possible what things are/mean.

jbygdell commented on scripts/sda-admin:
```diff
@@ -0,0 +1,448 @@
+
+myself=$0
+
+# Default values for global options.
+mq_credentials=test:test	# --mq-credentials user:pass
+mq_host=localhost		# --mq-host hostname
+sda_config=s3cmd.conf		# --sda-config pathname
+sda_key=crypt4gh_key.pub	# --sda-key pathname
```
Should these also be setable as ENVs?
It would make life easier when ingesting many files/dataset at the same time.

andkaha replied:
> Should these also be setable as ENVs? It would make life easier when ingesting many files/dataset at the same time.

Yes, this is easily done. I'll turn their names into uppercase at the same time to signal that they are initially taken from the environment if they are available there.

andkaha committed:
Address review comments from @jbygdell

* Make the whole MQ URL settable via command line option.
* Rename the MQ hostname variable `MQ_URL`.
* Also uppercase all user-settable variable names.
* Allow user-settable variables to be initialised from the environment.

andkaha committed:
Delay setting some variables

Some variables depend on the value of `$MQ_URL`, so delay setting these
until we know the final value of that variable.

andkaha committed:
Cleanup after renaming variable

norling commented on scripts/sda-admin:
```diff
@@ -0,0 +1,450 @@
+get_messages () {
+	# Retrieves the messages on the queue given by the 1st argument.
+	# The remaining arguments are filenames that we filter the
+	# messages with (togethur with the access key from the S3
```
```suggestion
	# messages with (together with the access key from the S3
```

norling commented on scripts/sda-admin:
```diff
@@ -0,0 +1,450 @@
+
+usage_accession () {
+	cat <<-USAGE_ACCESSION
+	The "accession" sub-command is used for assignig an accession ID
```
```suggestion
	The "accession" sub-command is used for assigning an accession ID
```

norling commented on scripts/sda-admin:
```diff
@@ -0,0 +1,450 @@
+
+	    $myself accession
+
+	    Accessions are only ever assiged to one file at a time.
```
```suggestion
	    Accessions are only ever assigned to one file at a time.
```

norling commented on scripts/sda-admin:
```diff
@@ -0,0 +1,450 @@
+url_exchanges=$url_api/exchanges/gdi/sda
+url_queues=$url_api/queues/gdi
+
+# Handle sub-comands.
```
```suggestion
# Handle sub-commands.
```

norling reviewed (approved):
Found a few typos, otherwise supernice stuff! :)

Commit:
Update scripts/sda-admin

Co-authored-by: Martin Norling <norlingjr@gmail.com>

Commit:
Update scripts/sda-admin

Co-authored-by: Martin Norling <norlingjr@gmail.com>

Commit:
Update scripts/sda-admin

Co-authored-by: Martin Norling <norlingjr@gmail.com>

Commit:
Update scripts/sda-admin

Co-authored-by: Martin Norling <norlingjr@gmail.com>

aaperis commented:
The script is a feast for the eyes :-). However, when I tried to run it a couple of things got in the way:

1. It seems to assume that the `sda-cli` binary is somewhere in my known path list. Maybe it would be useful to add a check somewhere in the beginning that all necessary tools are already installed.
2. Although this is not a problem of the script per se, the s3conf and c4gh pub key files are hidden inside a container volume so not directly accessible. 
I got them out using:
`docker cp ingest:/shared/c4gh.pub.pem .`
`docker cp ingest:/shared/s3cfg .`
and then tried the following:
```
./sda-admin --sda-config ./s3cfg --sda-key c4gh.pub.pem upload dummy_data
INFO[0000] Checking files
INFO[0000] Ready to encrypt 1 file(s)
INFO[0000] Reading Public key file                  
INFO[0000] Generating encryption key                    
INFO[0000] Encrypting file 1/1: dummy_data              
FATA[0000] failed to find credentials in configuration file
```
I think some documentation would be useful here.

Minor comment, repeating the above command did not printing any error which seems to be due to the 
`set -e` option. Better get an error than nothing though? :-)

andkaha commented:
> The script is a feast for the eyes :-). However, when I tried to run it a couple of things got in the way:
> 
>     1. It seems to assume that the `sda-cli` binary is somewhere in my known path list. Maybe it would be useful to add a check somewhere in the beginning that all necessary tools are already installed. 

`sda-cli` is only necessary for doing uploads. If you try to do an upload without that tool in your path, you get a standard error message pointing this out fairly promptly. 

> 
>     2. Although this is not a problem of the script per se, the s3conf and c4gh pub key files are hidden inside a container volume so not directly accessible.
>        I got them out using:
>        `docker cp ingest:/shared/c4gh.pub.pem .`
>        `docker cp ingest:/shared/s3cfg .`
>        and then tried the following:
> 
> 
> ```
> ./sda-admin --sda-config ./s3cfg --sda-key c4gh.pub.pem upload dummy_data
> INFO[0000] Checking files
> INFO[0000] Ready to encrypt 1 file(s)
> INFO[0000] Reading Public key file                  
> INFO[0000] Generating encryption key                    
> INFO[0000] Encrypting file 1/1: dummy_data              
> FATA[0000] failed to find credentials in configuration file
> ```
> 
> I think some documentation would be useful here.  

Someone (@jbygdell or @norling) said that there was a known(?) bug in `sda-cli`. If the configuration file has a `[default]` section header, it fails to read the configuration correctly.  This should not be documented for this tool but rather corrected in `sda-cli`.


> 
> Minor comment, repeating the above command did not print any error, which seems to be due to the `set -e` option. Better get an error than nothing, though? :-)

Using `set -e` does not suppress any error messages. I assume that the failed run nonetheless created c4gh files that the script then uploaded on the second attempt. There is no message upon successful uploads because the script does not check for success.  If you now try running `./sda-admin ingest`, you may possibly see your `dummydata` file being listed.  Also take a look at the container logs.

aaperis commented:
Oh yes, I forgot about the header issue of the `sda-cli`. Regardless, the point here is that the script is not testable without jumping through some hoops. But as I said, this has nothing to do with the script itself :-).

Now coming back to `set -e`, the script did not upload them in the second attempt since the problem with parsing the s3conf file still exists. Removing the `set -e` and repeating the same command as before returns the error as it should.

Edit: actually, `set -e` prevents the script from uploading in the case when there already exist `c4gh` files of the corresponding unencrypted files, regardless of the occurrence of any error. It  seems that in that case `encrypt ()` returns an `exit status` `1`.
Tried to call the script with `sh` in front in case it has something to do with my ubuntu terminal favoring bash but the behavior is the same.

andkaha commented:
> Oh yes, I forgot about the header issue of the `sda-cli`. Regardless, the point here is that the script is not testable without jumping through some hoops. But as I said, this has nothing to do with the script itself :-).
> 
> Now coming back to `set -e`, the script did not upload them in the second attempt since the problem with parsing the s3conf file still exists. Removing the `set -e` and repeating the same command as before returns the error as it should.
> 
> Edit: actually, `set -e` prevents the script from uploading in the case when there already exist `c4gh` files of the corresponding unencrypted files, regardless of the occurrence of any error. It seems that in that case `encrypt ()` returns an `exit status` `1`. Tried to call the script with `sh` in front in case it has something to do with my ubuntu terminal but the behavior is the same.

The `set -e` hid a bug introduced by https://github.com/GenomicDataInfrastructure/starter-kit-storage-and-interfaces/pull/4/commits/aeba14da121821c12ce94a52eb3e460c02c19636 yesterday.  I will correct that issue straight away. Basically, the script dies if there are no files to encrypt (because they are all already encrypted).

Question for you: Should files maybe be re-encrypted unconditionally?

andkaha committed:
Don't use short-circuit logic when `set -e` is in effect

aaperis commented:
> Question for you: Should files maybe be re-encrypted unconditionally?

Imo, unconditional re-encryption makes sense since this makes things simpler, helps that the user's files always are encrypted with the current key and I think the process is fairly fast either way.

andkaha commented:
> > Question for you: Should files maybe be re-encrypted unconditionally?
> 
> Imo, unconditional re-encryption makes sense since this makes things simpler, helps that the user's files always are encrypted with the current key and I think the process is fairly fast either way.

I will make it so.

aaperis commented on scripts/sda-admin:
```diff
@@ -0,0 +1,454 @@
+	shift 2
+done
+
+url_api=http://$MQ_URL:15672/api
```
```suggestion
url_api=$MQ_URL/api
```

andkaha replied:
Duh.

aaperis reviewed (changes requested):
small confusion here, I guess from recent changes. didnot go through the rest because I need to fix my c4gh pub key :-)

Edit: tested it till the end, looks and works great :-)

andkaha committed:
Unconditionally re-encrypt encrypted files

Commit:
Update scripts/sda-admin

Co-authored-by: Alex Aperis <76202622+aaperis@users.noreply.github.com>

andkaha commented:
> small confusion here, I guess from recent changes. didnot go through the rest because I need to fix my c4gh pub key :-)

The thing with the `url_api` and `MQ_URL` variables was definitely a bug caused by sloppy editing on my part. I committed your suggestion. Thanks!

andkaha committed:
Forgot to update documentation regarding unconditional encryption

jbygdell commented on scripts/sda-admin:
```diff
@@ -0,0 +1,459 @@
+MQ_CREDENTIALS=${MQ_CREDENTIALS-test:test}	# --mq-credentials user:pass
+MQ_URL=${MQ_URL-http://localhost:15672}		# --mq-url URL
+SDA_CONFIG=${SDA_CONFIG-s3cmd.conf}		# --sda-config pathname
+SDA_KEY=${SDA_KEY-crypt4gh_key.pub}		# --sda-key pathname
+
```
```suggestion
SDA_CLI=${SDA_CLI-sda-cli}                                      # path to sda-cli binary if not installed in $PATH

```

jbygdell commented on scripts/sda-admin:
```diff
@@ -0,0 +1,459 @@
+
+	# If there are files to encrypt, encrypt them.
+	if [ "$#" -gt 0 ]; then
+		sda-cli encrypt -key "$SDA_KEY" "$@"
```
```suggestion
		SDA_CLI encrypt -key "$SDA_KEY" "$@"
```

jbygdell commented on scripts/sda-admin:
```diff
@@ -0,0 +1,459 @@
+		set -- "$@" "$pathname"
+	done
+
+	sda-cli upload -config "$SDA_CONFIG" "$@"
```
```suggestion
	SDA_CLI upload -config "$SDA_CONFIG" "$@"
```

jbygdell reviewed (approved):
Might want to add some flexibility for people who don't want to put stuff in `$PATH`

andkaha commented:
> Might want to add some flexibility for people who don't want to put stuff in `$PATH`

This is a good idea. I'll add it with some comments in the code.

andkaha committed:
Allow user to point with etviroment variable `SDA_CLI` to `sda-cli` executable
